Clean up APIv3 error handling

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19048
## Description

This is an early start for improvements to the error handling in APIv3. I initially wanted to include these changes in another PR, but they are not needed there, so I postponed them to **this** PR, which will hopefully be finished in the coming weeks.
## Current content
- merging of model and contract errors moved into the contract
- removed secret side effect from error factory

myabc committed:
Remove livingstyleguide as Rails app dependency

NOTE: removes the ability to build the style guide with Sprockets.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Upgrade to livingstyleguide v2 syntax

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Rename style guide doc files from .md → .lsg

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add separate frontend Gemfile w/ livingstyleguide

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Bump gulp-livingstyleguide to 0.1.5

This version now uses `bundle exec` when calling out to the
`livingstyleguide` binary. See:
https://github.com/efacilitation/gulp-livingstyleguide/commit/d645828f591d453d15fac465b63956aa72882348

myabc committed:
[WORKAROUND] Add coffee-script as devDependency

Work around an issue with gulp-livingstyleguide packaging:
See https://github.com/efacilitation/gulp-livingstyleguide/issues/4

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add html[id] to style guide based on file path

Required in order for the style guide's own styles to be correctly
scoped. File path is most likely temporary however. See:
https://github.com/livingstyleguide/livingstyleguide/issues/133

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Tell gulp to watch .lsg files instead

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Commit:
Unified permission checks

Commit:
Bring activity_representer in line with the others

Commit:
Draft new work package create endpoints

Commit:
Fix activity user

Commit:
Swap work package creation links

Commit:
Extract shareable methods into separate module

Commit:
Increase service flexibility

Commit:
Introduce create work package endpoint

Commit:
Quieten the hound

Commit:
Collect writable attributes from ancestors

Commit:
Moved lock_version validation enforcement

Commit:
Provide more flexible structure for work package contract validations

Commit:
Appease the hound

Commit:
Introduce CreateWorkPackageService

Commit:
Improve initialization readability

Commit:
Move and rename AvailableWorkPackages API to meet standards

Commit:
Prepare WorkPackagesByProject for form endpoint

Commit:
Merge form folder into work_packages folder

Commit:
Improve refactorability

Commit:
Extract *_form_api endpoints

Commit:
Calm the hound

Commit:
Bump version to 4.3.0-alpha

Commit:
Accept projects as render context

NobodysNightmare committed:
Merge branch 'release/4.1' into dev

Commit:
Remove unnecessary destruction

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Separating form representer for update and create cases

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Consider contract information across the entire inheritance hierarchy

kgalli committed:
Merge branch 'release/4.2' into dev

Commit:
Adapt form representer tests

Commit:
Fix broken activity self_link

Commit:
Moved wp payload representer accordingly

Commit:
Some style fixes

Commit:
Fix permissions + cleanup

Commit:
Simplify work package services

NobodysNightmare committed:
extract error merging into contract

Commit:
Increase type safety ;)

NobodysNightmare committed:
remove side effect from Validation.create

I just don't know what the original author thought...

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-error-cleanup.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-error-cleanup.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
Make writable_attributes public

Commit:
Correct work_package_contract filename

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

0xF013 committed:
code extraction

0xF013 committed:
add missing Y

Commit:
Fix typo

Commit:
Remove erroneous existent project_id property

Commit:
Fix spec once again

floriank committed:
Merge pull request #3048 from 0xF013/feature/code-stabilization-for-datepickers

datepickers code stabilization

Commit:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Choose only valid types

Commit:
µstyle fix

ulferts commented on lib/api/contracts/model_contract.rb:
```diff
@@ -51,6 +51,27 @@ def self.attribute(*attributes, &block)
+        self.validate
+        model.valid?
+
+        errors = self.errors.dup
```
@NobodysNightmare We are in an instance, right? Wouldn't that mean that you take `errors` of the instance, dup them and then write the result back to `errors` of the instance. That would make it a NoOp. Haven't checked whether there is a setter for errors but even if what I say is wrong, you should employ a different name to avoid confusion.

NobodysNightmare replied:
I am pretty sure that the assignment reaches the local variable called errors.

But since you are asking the question in the first place, I see that there is room for confusion which I did not intend, so I'll rename it ^^

P.S.: I probably did not notice because of my code highlighting, which differenticates between local variables and instance methods.

ulferts commented on lib/api/contracts/model_contract.rb:
```diff
@@ -51,6 +51,27 @@ def self.attribute(*attributes, &block)
 
+      # Performs validation of both, the contract and the underlying model and returns
+      # the merged validation errors for both of them.
+      def validate_contract_and_model
```
I dislike the name here as you reveal to much of how you do it. I think you chose the name to avoid conflicting with the validate method that already exists but what is the drawback of overriding it?

NobodysNightmare replied:
Overriding is not an option as I also change the semantics (which is not even reflected in the new name).

`validate` performs validation, fills the `errors` and returns whether or not validation succeeded.

This method performs validation for both, contract and model and returns **the merged errors**, instead of a truthy/falsey value.

I wanted to depart from the practice of "polluting" the `errors` of either object with entries from the other one, that is why I do the duplication business. But this is also the reason I need to return the merged result immediately.

TBH: I already feel bad for calling the method `validate_x`, because you might not expect the return type I give to you. Which is more problematic in a dynamic language, because it is all about the expectations, isn't it? ^^

ulferts replied:
But why do you want to inform the sender of the method call of the complexity you have in the method. The sender shouldn't care whether you have to also ask the model whether it is valid. The sender should be able to ask whether the contract is fulfilled. 

Why not call it `violations` then?

NobodysNightmare replied:
So basically your idea is a different one than mine.
In my world contract and model are quite distinct. I just wanted to get this already existing merge-method to a more apropriate place and avoid that one `errors` object bleeds into the other one.

Your approach is different, you say "yes it is sad, that we have to validate both the contract and the model". But then you go on to say "but the caller does not need to know, we make him believe he only validates the contract".
This is vastly different from what I came up with, but I tell you: **_I LIKE IT!**_

My course of action would now be to:
- call this baby `validate` and make it behave like the `super.validate` w.r.t return values and stuff
- merge the errors of the model and the contract into the _contract errors_
- let the caller believe it is only validating the contract alone

Benefits being that as soon as the day arrives when we validate via contracts only, we just need to remove this sweet little method override.

Do you agree? Did I understand you correctly?

NobodysNightmare committed:
rename local

Commit:
Improve shared helper encapsulation

Commit:
Tiny style fix

ulferts commented on lib/api/errors/validation.rb:
```diff
@@ -53,19 +53,21 @@ def self.create(errors)
-            errors.set(key, error)
-            errors.delete(p)
+            merged_errors.set(key, error)
+            merged_errors.delete(property)
```
Hm, somebody forgot to write specs for the whole class. As you fixed an existing bug, I think you should at least write specs for `merge_error_properties`.

ulferts commented:
@NobodysNightmare overall, I like the direction this is heading.

NobodysNightmare committed:
hide error merging behind contract.validate

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

Commit:
Add all the specs

Commit:
Remove unused property

Commit:
Clarify comment

Commit:
Add condition to else

Commit:
Unify activity path declaration

Commit:
Use implicit getter

Commit:
Shorten form_representer by inheriting from single

Commit:
Add complimentary post

Commit:
Remove duplicated permission check

Commit:
Move service creation into post

Commit:
Remove superfluous variable

Commit:
Style fixes

Commit:
Correct wp create endpoint refactoring

Commit:
Adjust spec to service change

Commit:
Lull the hound

Commit:
Remove activity self link title

Commit:
Use proper inherited variable

Commit:
Move helper method to helpers

Commit:
Add specs for create form api

Commit:
Explicitly specify PayloadRepresenter

Commit:
Fix dumb mistake

Commit:
Remove outdated test

NobodysNightmare committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Properly use caching mechanism

Commit:
Remove debug remnant

Commit:
Diversify tests

Commit:
Remove model_ prefix

Commit:
Cleanup test

Commit:
Clarify spec purpose

Commit:
Improve test setup and checks

Commit:
Ditch regex for more readable string

Commit:
Check link content as well

Commit:
Move test to proper location and benefit from it

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Refactor test to use helper_class

Commit:
Simplify test setup

Commit:
Reduce code duplication

Commit:
Lower expectations

Commit:
Add tests for notification case

Commit:
Diversify tests

kgalli committed:
Merge branch 'release/4.2' into dev

Commit:
Move out of place parenthesis back into position

Commit:
pin select 3.5.2

Commit:
Add test descriptions

Commit:
Correct test description

Commit:
Remove redundant context

Commit:
Remove temporal coupling

ulferts committed:
Merge pull request #2990 from Azure7111/feature/create_work_package

Add create work package endpoints

Commit:
Allow raising exceptions from notification jobs

machisuji committed:
Merge pull request #3126 from barhot/feature/allow-failing-notification-jobs

Allow raising exceptions from notification jobs

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

myabc committed:
Bump rspec, rspec-rails to latest ~> 3.3.0

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix running of specs with rspec executable

Legacy specs should be filtered out by default when using `rspec`
executable with no arguments or a path that contains both legacy and
non-legacy specs (e.g. `rspec spec`).

As noted in 1a0e2b90, suites must be run separately for time-being.

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts committed:
Merge pull request #3143 from myabc/fix/specs-rspec-executable

Fix running of specs with rspec executable

Commit:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge pull request #3142 from myabc/feature/rspec-3.3

RSpec 3.3 Upgrade

ulferts committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
Merge pull request #3120 from floriank/fix/pin-select-version

[18714] Pin select2 3.5.2

NobodysNightmare committed:
Merge pull request #2957 from myabc/feature/livingstyleguide-upgrade

ulferts committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

ulferts committed:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

Commit:
Merge branch 'release/4.2' into dev

NobodysNightmare committed:
fix stupid mistakes

- reference to method by old name
- not knowing how to call a super-method

NobodysNightmare committed:
dirty trick to allow using a Hash as mock

also removes default value from initializer, which was never sane
